be/src/exprs/time_functions.cpp

```diff
@@ -2361,4 +2361,125 @@ StatusOr<ColumnPtr> TimeFunctions::make_date(FunctionContext* context, const Col
+
+    ColumnBuilder<TYPE_DATE> result(size);
+    for (int row = 0; row < size; ++row) {
+        if (data_column.is_null(row)) {
+            result.append_null();
+            continue;
+        }
+
+        DateValue date = (DateValue)data_column.value(row);
+        if (format_content == "year") {
+            date.set_end_of_year();
+        } else if (format_content == "quarter") {
+            date.set_end_of_quarter();
+        } else {
+            date.set_end_of_month();
+        }
+        result.append(date);
+    }
+    return result.build(ColumnHelper::is_all_const(columns));
+}
+
+StatusOr<ColumnPtr> TimeFunctions::_last_day_with_format_optional(FunctionContext* context, const Columns& columns) {
+    ColumnViewer<TYPE_DATETIME> data_column(columns[0]);
+    ColumnViewer<TYPE_VARCHAR> format_column(columns[1]);
+    auto size = columns[0]->size();
+
+    ColumnBuilder<TYPE_DATE> result(size);
+    for (int row = 0; row < size; ++row) {
+        if (data_column.is_null(row) || format_column.is_null(row)) {
+            result.append_null();
+            continue;
+        }
+
+        DateValue date = (DateValue)data_column.value(row);
+        Slice format_content = format_column.value(row);
+        if (!(format_content == "year" || format_content == "month" || format_content == "quarter")) {
+            Status::InvalidArgument("last day optional in {month, quarter, year}, but optional is " +
+                                    format_content.to_string());
+        }
+
+        if (format_content == "year") {
+            date.set_end_of_year();
+        } else if (format_content == "quarter") {
+            date.set_end_of_quarter();
+        } else {
+            date.set_end_of_month();
+        }
+        result.append(date);
+    }
+    return result.build(ColumnHelper::is_all_const(columns));
+}
+
+StatusOr<ColumnPtr> TimeFunctions::last_day_with_format(FunctionContext* context, const Columns& columns) {
+    DCHECK_EQ(columns.size(), 2);
+    auto* state = reinterpret_cast<LastDayCtx*>(context->get_function_state(FunctionContext::FRAGMENT_LOCAL));
+    if (state->const_optional) {
+        std::string format_content = state->optional_content;
+        return _last_day_with_format_const(format_content, context, columns);
+    }
+    return _last_day_with_format_optional(context, columns);
+}
+
+Status TimeFunctions::last_day_prepare(FunctionContext* context, FunctionContext::FunctionStateScope scope) {
+    if (scope != FunctionContext::FRAGMENT_LOCAL) {
+        return Status::OK();
+    }
+
+    auto* state = new LastDayCtx();
+    context->set_function_state(scope, state);
+
+    if (!context->is_notnull_constant_column(1)) {
+        return Status::OK();
+    }
+
+    state->const_optional = true;
+    ColumnPtr column = context->get_constant_column(1);
+    Slice optional = ColumnHelper::get_const_value<TYPE_VARCHAR>(column);
+    if (!(optional == "year" || optional == "month" || optional == "quarter")) {
```

Comment:
case-sensitive?  can we use Year, MoNth, quartER?

Reply:
case-sensitive, only "year"、"month"、"quarter"